src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs

```diff
@@ -126,19 +136,19 @@ public CompilerCache(IFileProvider fileProvider)
 
                 cacheEntryOptions = new MemoryCacheEntryOptions();
 
-                compilerCacheContext = cacheContextFactory(normalizedPath);
-                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(compilerCacheContext.ProjectItem.Path));
-                if (!compilerCacheContext.ProjectItem.Exists)
+                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(normalizedPath));
+                var projectItem = _templateEngine.Project.GetItem(normalizedPath);
```

Comment:
I'm not pushing back on the use of the template engine, I'm just suggesting maybe it's not a **direct** relationship